### ISSUE TITLE
certificate name without trailing dot for avahi hosts

### ISSUE THREAD

**Issue by fanto666** [mumble-voip/mumble#5268]
**Describe the bug**
when connecthing to host "xxxxe.local" found using avahi on local network, mumble connects to "xxxxe.local." - name ending with trailing dot, certificate for "xxxxe.local" (without trailing dot) is refused.

Creating certificate with SubjectAltName containing trailing dot does work around the problem. 

**Steps to Reproduce**
- run mumble
- connect to host on local network
- you'll see that hostname mumble is connecting to contains trailing dot

**Expected behavior**
trailing dot should not be used in certificate check.

**Screenshots**
![mumble2](https://user-images.githubusercontent.com/483529/133937883-6aa97274-becc-4958-a1c1-4cfe1cd4eaae.png)

**Desktop (please complete the following information):**
 - OS: Debian GNU/Linux 11, amd64
 - Version: 1.3.4

**Additional context**

**Comment by Krzmbrzl** [comment]
So it seems the underlying issue is that Mumble uses a hostname ending in a trailing dot (for all servers in local network).

This is also reproducable when hosting a server on the same machine as the client is running (using Linux).

**Comment by Krzmbrzl** [comment]
From a quick look at the sources, it seems that this is a Zeroconf artifact:
https://github.com/mumble-voip/mumble/blob/296956c3e3e6100f05f6f2d1525fd592ef567a17/src/mumble/ConnectDialog.cpp#L221
https://github.com/mumble-voip/mumble/blob/296956c3e3e6100f05f6f2d1525fd592ef567a17/src/mumble/ConnectDialog.cpp#L262
(4 more)

@davidebeatrici Any specific reason you are using the trailing dot here?

**Comment by fanto666** [comment]
I guess it is to prevent appending domain name in DNS searchlist to the locally detected hostname.

**Comment by davidebeatrici** [comment]
If I recall correctly it's because the Windows API fails if there is no trailing dot.

**Comment by Krzmbrzl** [comment]
Okay so according to https://serverfault.com/questions/18113/dns-trailing-periods the trailing period is required in order to turn this into a fully qualified domain name (absolute). Without the trailing dot it seems the domain name is only a relative one.

If what is written in the linked post is correct it seems that Mumble correctly infers the origin for the relative address to be the current origin `.` and by appending it, it turns the relative domain name into an absolute one.
Thus under these assumptions it seems that Mumble is actually doing the right thing here :thinking: 

Now you wrote that this is causing issues with the used certificate. What kind of certificate is that and how was it created? Is it an auto-generated cert created by Mumble or something else?

**Comment by fanto666** [comment]
it's certificate created by personal CA for machine running (not only) murmur server.
the CA is installed on local network so it could be trusted by mumble client, but first I needed to recreate certificate to add Subject Alternative Name with trailing dot.

I know that trailing dots are used in DNS to distinguish FQDN from relative name, however they are not used in certificates - I guess names in certificates are explicitly fully qualified.

Is it possible to strip trailing dot from the resolved .local name before comparing certificates?

**Comment by davidebeatrici** [comment]
Yes, absolutely. Is it the only TLD we should apply the exception to?

**Comment by Krzmbrzl** [comment]
I think removing the trailing dot is not the correct solution. Instead we should figure out why a FQDN does fail to verify. What is the exact error that you are getting when connecting to your server? And what does the server log have to say about that connection?

**Comment by fanto666** [comment]
Since all FQDN DNS records end with dot (reason described above) but the dot is stripped in clients, and because certificates don't contain names with trailing dots (I have never seen any, and I maintain servers with letsencrypt and commercial certificates), I believe stripping trailing dot is the correct way to solve this problem.

I don't think server log is important, because it's the client that complains about server certificate not matching required hostname and not talking to the server before user confirms that the server is okay.
I have re-created the certificate without trailing dot in SAN:
![cert-err1](https://user-images.githubusercontent.com/483529/134800990-7fe6b7c1-d586-4ca8-8967-69977f5135c7.png)
![cert-err2](https://user-images.githubusercontent.com/483529/134800992-5780c9fd-635e-4a23-8e68-f1f871c9b873.png)
if I add fourth record with trailing dot, DNS:*.local. the connection is OK
![cert-err3](https://user-images.githubusercontent.com/483529/134801181-6258aa32-4d74-48ee-be69-6fd292ec12a4.png)
